File: framework/test2/common/intbreak/intbreak.rb
# step [1]: implement solution function
# @param {Integer} n
# @return {Integer}
def integer_break(n)
  solve(n)
end

$memo = Array.new(59, 0)

def solve(n)
  if n == 1
    return 1
  end
  if $memo[n] > 0
    return $memo[n]
  end
  res = 0
  1.step(n-1) { |i|
    res = [i*(n-i), i*solve(n-i), res].max
  }
  $memo[n] = res
  res
end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  work = ns::TestWork.create(method(:integer_break))
  # work.validator = -> (e, r) { e is equal to r }
  work.compareSerial = true
  puts work.run(ARGF.read)
end

File: framework/test2/common/intbreak/intbreak.pl
use strict;
use warnings;
use List::Util qw(min max);

# step [1]: implement solution function
# @param {number} n
# @return {number}
sub integer_break {
    my ($n) = @_;
    solve($n);
}

my @memo = (0) x 59;

sub solve {
    my ($n) = @_;
    if ($n == 1) {
        return 1;
    }
    if ($memo[$n] > 0) {
        return $memo[$n];
    }
    my $res = 0;
    for my $i (1..$n-1) {
        $res = max($i * ($n-$i), $i * solve($n-$i), $res);
    }
    $memo[$n] = $res;
}

use Soda::Unittest::Utils qw(from_stdin);
use Soda::Unittest::TestWork qw(create);
# step [2]: setup function/return/arguments
my $work = create(\&integer_break);
# $work->validator(sub { $_[0] vs $_[1]; return boolean });
$work->compare_serial(1);
print $work->run(from_stdin());

